### ISSUE TITLE
netmd download

### ISSUE THREAD

**Issue by Lt-Henry** [linux-minidisc/linux-minidisc#18]
I've been trying to get netmd track download to work, but had some troubles and doubts. First of all, why is LP2 wireformat default? There is no way to encode atrac in Linux, mp3, ogg and flac are more common and easily re-encoded to raw pcm wireformat. I've been working this way with the python version for a long time.

I've found a pair of bugs in secure.c (or that seems to me). 

Line 394: size_t chunksize = 0xffffffffU;

Should be chunk size 4Gb? libusb_bulk_transfer fails when trying to send the whole track in one single packet (~48Mb). Smaller values work for me, but not sure how to determine the best value.

Line 384: There is a lonely break which only allows to one packet to be sent. I guess a "else" scope is missing.

Fixing both lines sent the whole track down to the MD, but after that, it fails polling for a response (Still guessing why :S ) and the sync toc and so one seems to have no effect, device gets stuck "accessing" forever.

Tested on Ubuntu 16.04 x86_64

**Comment by glaubitz** [comment]
> I've been trying to get netmd track download to work, but had some troubles and doubts. First of all, why is LP2 wireformat default? There is no way to encode atrac in Linux, mp3, ogg and flac are more common and easily re-encoded to raw pcm wireformat. I've been working this way with the python version for a long time.

It's been a long time, but I was pretty sure you could only download ATRAC-SP or ATRAC-LP to most NetMD devices but @karcherm knows more on this and he might be able to answer.

I assume you were using _netmdcli_ for downloading. Looking at the code, **netmdcli** currently hardwires the format to LP2 but that should be easy to fix in the future.

> I've found a pair of bugs in secure.c (or that seems to me).

You should open separate bugs for that. Logically separate issues should not thrown together into single bug report.

> Tested on Ubuntu 16.04 x86_64

The version of the operating system is not really relevant but the version of linux-minidisc is. Please report issues against the current git version only.

**Comment by tharp** [comment]
> Fixing both lines sent the whole track down to the MD, but after that, it fails polling for a response (Still guessing why :S ) and the sync toc and so one seems to have no effect, device gets stuck "accessing" forever.

Device gets stuck showing "accessing" when the 
data size to be transferred doesn't match the data size (total_bytes) send to the device at netmd_secure_send_track() command. In some cases the device is still waiting for more audio data, so it stucks and blocks any other command if it doesn't get the "missing data".
It's just a matter of wrong/incomplete calculation. ).
The last packet/chunk may be limited in size (secure.c line 417) and maybe frames will be truncated, but this will not be recognized when calculating the total bytes to be sent (secure.c line 513).

OT: I made some changes some time ago by calculating the number of frames stored in the packets (when i worked on integrating netmd support in qhimdtransfer).
It's not reviewed (i didn't sent further pull requests) and also needs some work. Made some first steps for unix autodetection, too. 
I'm very busy now but if anyone wants to test/review it or continue working on it feel free to take a look at my "unixauto" branch. (it is based on an older version, so it has to be rebased, download of Atrac LP and PCM is supported)

**Comment by Lt-Henry** [comment]
> It's just a matter of wrong/incomplete calculation. ).

Yes, it was! I've already got it working. Netmdcli code was skiping a 60 bytes header, that my raw pcm sources lacks, so total transfer was smaller than expected. Thanks @tharp !

> The last packet/chunk may be limited in size (secure.c line 417) and maybe frames will be truncated, but this will not be recognized when calculating the total bytes to be sent (secure.c line 513).

I am still a bit confused about those PCM not multiple of 2048, so this last frame is lost?

> I'm very busy now but if anyone wants to test/review it or continue working on it feel free to take a look at my "unixauto" branch

I will take a look!

> You should open separate bugs for that. Logically separate issues should not thrown together into single bug report.

Sorry @glaubitz I wasn't sure if using lp2 as default format could be considered as a bug, also, I don't have a clean solution for that now, other than hardwiring for PCM input. I am going to clean up the code and submit a pull request for the other.

**Comment by tharp** [comment]
> I am still a bit confused about those PCM not multiple of 2048, so this last frame is lost?

O.K., this should not happen, maybe i have had some bad/corrupt wav files for testing. I implemented counting frames for debug messages and then kept it for calculating transfer data size. Anyway, this worked fine for me so far.
If PCM audio data size is not multiple of 2048 (should not happen on correct files) the last frame is not lost but may be shorter than expected, so the device sticks waiting for the "missing data".

**Comment by Lt-Henry** [comment]
> If PCM audio data size is not multiple of 2048 (should not happen on correct files)

I'm converting from mp3 to pcm using this:
ffmpeg -i track.mp3 -f s16be -acodec pcm_s16be track.raw

The output size for this file is 49844736, not a 2048 multiple. This happens with almost file tested. The hexdump does not show any header. Maybe i'm doing something wrong :s  

> the last frame is not lost but may be shorter than expected, so the device sticks waiting for the "missing data".

After removing fseek of 60 bytes, this never happen again. I downloaded a pair of albums without problems. I guess the magic is in the integer division 49844736÷2048=24338,25 -> 24338 so the downloaded track is missing 512 bytes of track (~ 3ms). But this is a fast guess.

**Comment by tharp** [comment]
> The hexdump does not show any header. Maybe i'm doing something wrong :s 

There is nothing "really" wrong, you are converting your mp3 file to pcm raw audio. There is no header then but the file is useless for anyone who doesn't know the audio format.
I prefer converting to a wav file (pcm audio stored in a wav container), it contains the header storing all information about the audio format (bitrate, channels, sample rate etc).
You can do this by calling ffmpeg whithout any options and using .wav as output file extension. (" ffmpeg -i track.mp3 track.wav").
ffmpeg uses default values for the output file which is 16bit pcm @ 44k1 Hz, stereo.

> After removing fseek of 60 bytes, this never happen again.

O.K., this "fseek of 60 bytes"-section is for skipping the header, but it is just a workaround and only works on files with a header length of 60 byte.
This is what i changed in "my unixauto branch" by reading the header, setting the correct transfer options according to the audio format etc.

>   output size for this file is 49844736, not a 2048 multiple. This happens with almost file tested

O.K., the function preparing the audio packets don't take care of frames and just uses the existing audio data  but the calculation of the total amount of bytes to transfer uses frames.
There are 2 ways to fix this:
1. take care of frames when preparing the packets by padding the last "incomplete" frame with 0x00
2. don't take care of frames when calculating total amount of bytes, use existing audio data size instead

i prefer the 1st one because i think it's "easier/better" for the netmd device coding the pcm audio to Atrac if frames are "complete"

**Comment by Lt-Henry** [comment]
Ok, so It is clear now where the magic number 2048 comes from.

> This is what i changed in "my unixauto branch" by reading the header, setting the correct transfer options according to the audio format etc.

Yes, I took a look at your branch. Would be nice to get it working again, because current master branch is totally useless for netmd devices. My local branch is working with raw pcm as you may already guess :) 

IMHO, ffmpeg is the way to go. It is a well known and portable library, it will allow us to accept almost everything as input... everything but Atrac :_D in that case we can perform a 'direct' download.

> 1. take care of frames when preparing the packets by padding the last "incomplete" frame with 0x00

For sure  this is the easy way, but correct me if I am wrong, does this extra padding inject an extra silence to the track, up to 2044 bytes? ~10ms

**Comment by glaubitz** [comment]
So, I can't comment on this directly but @karcherm can. Please wait for him to have a look at this and your PR.

**Comment by tharp** [comment]
> Line 394: size_t chunksize = 0xffffffffU;
> Should be chunk size 4Gb? libusb_bulk_transfer fails when trying to send the whole track in one single packet (~48Mb). Smaller values work for me, but not sure how to determine the best value.

I don't remember exactly but afair from my tests "real" netmd devices (not himd devices in netmd mode) don't work with multiple packets, so increasing libusb timeout works for me.
Transferring tracks on himd devices in netmd mode work fine with multiple packets and smaller chunk size.
Maybe someone can make some more tests on this?

**Comment by Lt-Henry** [comment]
> Maybe someone can make some more tests on this?

I've been reading libusb docs and talking with @karcherm about that. The way usb transfers are done is buggy. There is no need to split into packets, as libusb already does it. Timeouts should be handled properly, because not always mean something wrong happened. I am working on a fix for that. 

I think this issue can be closed as we already know what fails and how to fix it.
